Lex persona gateway (#349)
openfun/joanie: docs/explanation/lex-persona.md

This is the documentation on how to implement Lex Persona Workflow Manager, which is related to the issue : https://github.com/openfun/joanie/issues/318

> Developer's Documentation : https://sgs-demo-test01.sunnystamp.com/wm-docs/index.html

Author: You are right, I should remove it right away.  My bad.

            "maxInvites": 5,
            "recipients": [
                {
                    "consentPageId": "cop_NzBEGbgTkj7v5tq********",
                    "country": "FR",
                    "email": "<signer_email_address>@mail.com",
                    "firstName": "John",
                    "lastName": "Doe",
                    "phoneNumber": "+33 7 ** ** ** **",
                    "preferredLocale": "fr"
                }
            ],
            "requiredRecipients": 1,
            "sendDownloadLink": true,
            "stepType": "signature",
            "validityPeriod": 8553600000
        }
    ],
    "tenantId": "ten_9H4TvNNQFfrLBXH********",
    "updated": 1688993660520,
    "userId": "usr_P5YBCEeRqY8jKGG********",
    "viewAuthorizedGroups": [
        "grp_9woWtcSt5KiAzyq********"
    ],
    "viewAuthorizedUsers": [],
    "watchers": [],
    "workflowStatus": "stopped"
}
```
Once, this is done, you have just created the workflow, well done !

In this example, I've decided to send the document to be signed to a user that is not yet registered on the Lex Persona BackOffice.

This example shows you that it is not required to create a "user".

> We can just parse the signer's details into the "recipients" lists of the payload (firstName, lastName, phoneNumber, email).

Author: **When the signer is an external contact in LexPersona :** 

It is required to send the `firstName` and the `lastName` of the signer if she/he is an external contact.  If one of those fields is missing, the API responds with a Bad request - 400 : "A recipient in the request is missing identity information." It also responds with a Bad request if the `email` is missing.

If we provide a `userId`, we do not need to send the `firstName`, `lastName`, `email` and `phoneNumber` since this data is saved into the user's object (this example requires to create a user for each signer).

If we do not provide a `userId` , the required data of a contact are the following ones: 
- With SMS validation : 
    - `firstName`, `lastName`, `email`, `phoneNumber`

- Without SMS validation : 
    - the `phoneNumber` is not required.

> We can just parse the signer's details into the "recipients" lists of the payload (firstName, lastName, phoneNumber, email).

Author: Info : If we only put the `name` as a full name in the payload instead of `firstName` and `lastName`, the API responds with a 400 - Bad Request.

            "logs": [
                {
                    "created": 1688994896088,
                    "operation": "start"
                }
            ],
            "maxInvites": 5,
            "recipients": [
                {
                    "consentPageId": "cop_NzBEGbgTkj7v5tq********",
                    "country": "FR",
                    "email": "<signer_email_address>@mail.com",
                    "firstName": "John",
                    "lastName": "Doe",
                    "phoneNumber": "+33 7 ** ** ** **",
                    "preferredLocale": "fr"
                }
            ],
            "requiredRecipients": 1,
            "sendDownloadLink": true,
            "stepType": "signature",
            "validityPeriod": 8553600000
        }
    ],
    "tenantId": "ten_9H4TvNNQFfrLBXH********",
    "updated": 1688994896088,
    "userId": "usr_P5YBCEeRqY8jKGG********",
    "viewAuthorizedGroups": [
        "grp_9woWtcSt5KiAzyq********"
    ],
    "viewAuthorizedUsers": [],
    "watchers": [],
    "workflowStatus": "started"
}
```
We can observe that the "workflowStatus" value is set to "started" and the "progress" is set to 0.

> At this point, the recipients will receive through email a notification that a document requires to be signed by them.

Reviewer: Is there a way to not send an email but just retrieve the link to sign the document and display it to the user on a UI ?

Author: Indeed, there is a way to create the invitation link which will allow us to display it to the user on UI.
The endpoint is : /api/workflow/{{id_workflow}}/invite.

The way is to set it up when we create the workflow, where we set to the `maxInvites` field the value 0.

___

### III) Signing the document (user's view on his browser)

#### 1- The signer side

![Alt text](../assets/lex_persona_schema_user.png)

At the end of this step, meaning that the user has signed the document, we will receive
an email notification suggesting the signers to download the freshly signed documents.

#### 2- Retrieve the status of the workflow

Once those steps through the user are finished, we can now check our workflowStatus :

**Endpoint** : /api/workflows/{{workflow_id_created}}

**Method** : GET

**Response**:
```json

> {
>     "created": 1689089045598,
>     "currentRecipientEmails": [],
>     "currentRecipientUsers": [],
>     "description": "91 Allée, Voie de Presbourg, 60282 Vénissieux",
>     "email": "<user_email>@mail.fr",
>     "firstName": "Sam",
>     "groupId": "grp_9woWtcSt5KiAzyq********",
>     "id": "wfl_A1pRnmnXfW2G2akS********",
>     "lastName": "P******",
>     "logs": [],
>     "name": "Aerodynamic Silk Bottle",
>     "notifiedEvents": [
>         "recipientRefused",
>         "recipientFinished",
>         "workflowStopped",
>         "workflowFinished"
>     ],
>     "progress": 100,
>     "started": 1689092760018,
>     "steps": [

Reviewer: Ditto

```suggestion
{
    "created": 1689089045598,
    "currentRecipientEmails": [],
    "currentRecipientUsers": [],
    "description": "1 rue de l'exemple, 75000 Paris",
    "email": "johndoe@example.com",
    "firstName": "John",
    "groupId": "grp_9woWtcSt5KiAzyq********",
    "id": "wfl_A1pRnmnXfW2G2akS********",
    "lastName": "Doe",
    "logs": [],
    "name": "Aerodynamic Silk Bottle",
    "notifiedEvents": [
        "recipientRefused",
        "recipientFinished",
        "workflowStopped",
        "workflowFinished"
    ],
    "progress": 100,
    "started": 1689092760018,
    "steps": [
```

                },
                {
                    "created": 1689092977709,
                    "operation": "notifyWorkflowFinished"
                }
            ],
            "maxInvites": 5,
            "recipients": [
                {
                    "consentPageId": "cop_NzBEGbgTkj7v5tqp****",
                    "country": "FR",
                    "email": "<signer_email_address>@mail.com",
                    "firstName": "John",
                    "lastName": "Doe",
                    "phoneNumber": "+33 7 ** ** ** **",
                    "preferredLocale": "fr",
                    "userId": "usr_H5RSBaPTeYu7********"
                }
            ],
            "requiredRecipients": 1,
            "sendDownloadLink": true,
            "stepType": "signature",
            "validityPeriod": 8553600000
        }
    ],
    "tenantId": "ten_9H4TvNNQFfrLBXH2********",
    "updated": 1689092977939,
    "userId": "usr_P5YBCEeRqY8jKGGM********",
    "viewAuthorizedGroups": [
        "grp_9woWtcSt5KiAzyqL********"
    ],
    "viewAuthorizedUsers": [
        "usr_H5RSBaPTeYu7EjsG********"
    ],
    "watchers": [],
    "workflowStatus": "finished"
}
```

> We can see the workflowStatus has reached his final step : "finished"

Reviewer: ```suggestion
We can see the `workflowStatus` has reached his final step : "finished"
```

___
#### 4- Retrieve the signed documents :

For a normal signer who was invited to sign a document, he will be able to retrieve the documents through the last notification email that states the end of the worklfow.

> Concerning the API, we can just use the API to retrieve the documents that were signed.

Reviewer: How long is the document stored by Lex Persona? Should we have to store it once signed on our own infrastructure or can we deal with the fact that Lex Persona document life time is long enough ?

Author: Lex Persona stores your document from 6 years up to 10 years. Documents can be accessible for this length of time.

## API Tutorial - Workflow Manager
As we have mentioned in the first section of this document, there are a few API calls (5) to execute in order to launch the workflow and get back the document fully signed.

### REQUIREMENT
You **MUST** have a **Bearer Token**, that is kept into the .env file (or else where depending on your setup) of your project. This key **MUST** remain secret.

Once we will have implemented our LexPersona gateway, we will update this part to mention which environment variable the developer should set for this purpose.

### SCHEMA
1. Structured data are sent and received as JSON
2. All timestamps are returned in milliseconds since January 1, 1970.

___

### Here are the main steps :
![schema describing the main steps of the API with lex persona](../assets/lex_persona_schema_api.png)

To summarize the main steps with the API :

- Create the workflow (POST method)
- Upload the document to be signed (POST method)
- Configure the signature zones (POST method)
- Update the `workflowStatus` of the workflow that we have just created to "started" (PATCH method)
- Retrieve the signed documents (GET method)

> Additionnaly, you can create the invitation link with a POST request in order to have the control on how we will deliver this link to our future signer. This demands to set the `maxInvites` to 0 when we create the workflow in the payload, which gives us the control for the invitation link overtime (more details further down the documentation).

Reviewer: ```suggestion
Additionnaly, you can create the invitation link with a POST request in order to have the control on how we will deliver this link to our future signer. This implies to set the `maxInvites` to 0 when we create the workflow in the payload, which gives us the control for the invitation link overtime (more details further down the documentation).
```

            "maxInvites": 5,
            "recipients": [
                {
                    "consentPageId": "cop_NzBEGbgTkj7v5tq********",
                    "country": "FR",
                    "email": "johndoe@example.com",
                    "firstName": "John",
                    "lastName": "Doe",
                    "phoneNumber": "+33 * ** ** ** **",
                    "preferredLocale": "fr"
                }
            ],
            "requiredRecipients": 1,
            "sendDownloadLink": true,
            "stepType": "signature",
            "validityPeriod": 8553600000
        }
    ],
    "tenantId": "ten_9H4TvNNQFfrLBXH********",
    "updated": 1688993660520,
    "userId": "usr_P5YBCEeRqY8jKGG********",
    "viewAuthorizedGroups": [
        "grp_9woWtcSt5KiAzyq********"
    ],
    "viewAuthorizedUsers": [],
    "watchers": [],
    "workflowStatus": "stopped"
}
```
Once this is done, you have just created the workflow, well done !

In this example, I've decided to send the document to be signed to a user that is not yet registered on the Lex Persona BackOffice. Let's dive into more details **when the signer is an external contact in LexPersona :**

> It is required to send in the request's payload the `firstName` and the `lastName` of the signer if she/he is an external contact. If one of those fields is missing, the API responds with a Bad request - 400 : "A recipient in the request is missing identity information." It also responds with a Bad request if the `email` is missing.

Reviewer: In english, we can use `it` or `they` for neutral gender.
```suggestion
It is required to send in the request's payload the `firstName` and the `lastName` of the signer if it is an external contact. If one of those fields is missing, the API responds with a Bad request - 400 : "A recipient in the request is missing identity information." It also responds with a Bad request if the `email` is missing.
```

When we receive the response of our POST request to create the workflow, we can see that the `workflowStatus` as the value **"stopped"**. Which indicates that we will need in another request to **start** the **workflow** process.

```json
"workflowStatus": "stopped"
```
But first, we **must** add (upload) the document to be signed in the workflow that has been just created.
___
#### 2 - Uploading the document to be signed
**Endpoint** : api/workflows/{{workflow_id_created}}/parts?createDocuments=true&signatureProfileId={{signature_profile}}

**Method** : POST

**Headers**:
- Authorization : Bearer {{token}}
- Content-Type : multipart/form-data

**Params** :
- key: createDocument,
value: true
- key:  signatureProfileId,
value: sip_…

**Body Type** : form-data
- key: file1, value: le_fichier_pdf_a_soumettre.pdf

**Response** :
```json
{
    "documents": [
        {
            "created": 1688994064535,
            "groupId": "grp_9woWtcSt5KiAzyq********",
            "id": "doc_JyjUX6uyzFBcjTx********",
            "parts": [
                {
                    "contentType": "application/pdf",

>                     "filename": "le_fichier_pdf_a_soumettre.pdf",

Reviewer: use english please.

```suggestion
                    "filename": "the_file_to_sign.pdf",
```

**B) Hash signature** :

Once the document's hash has been created, it is signed using the private key of a pair of cryptographic keys (public key / private key). The private key is owned only by the person signing the document, while the public key can be shared with other users.

**C) Add the signature to the document** :

The resulting signature (which is essentially a string of bits) is added to the original document, creating the electronically signed version.

**D) Signature verification** :

If someone wishes to check the validity of the document, the process goes the other way around, which means : the signer extracts the signed document's hash and he uses the public key in order to verify if the hash's signature is valid.

If the verification succeeds, it means that the document was not altered (changed) and the signature was performed by a person holding the private key corresponding to the public key used for verification.

> Globally, the verification of signature's hash can be made with Adode.

Reviewer: ```suggestion
Globally, the verification of signature's hash can be made easily through Adode product.
```

            "isStarted": true,
            "logs": [
                {
                    "created": 1688994896088,
                    "operation": "start"
                }
            ],
            "maxInvites": 5,
            "recipients": [
                {
                    "consentPageId": "cop_NzBEGbgTkj7v5tq********",
                    "country": "FR",
                    "email": "johdoe@example.com",
                    "firstName": "John",
                    "lastName": "Doe",
                    "phoneNumber": "+33 * ** ** ** **",
                    "preferredLocale": "fr"
                }
            ],
            "requiredRecipients": 1,
            "sendDownloadLink": true,
            "stepType": "signature",
            "validityPeriod": 8553600000
        }
    ],
    "tenantId": "ten_9H4TvNNQFfrLBXH********",
    "updated": 1688994896088,
    "userId": "usr_P5YBCEeRqY8jKGG********",
    "viewAuthorizedGroups": [
        "grp_9woWtcSt5KiAzyq********"
    ],
    "viewAuthorizedUsers": [],
    "watchers": [],
    "workflowStatus": "started"
}
```
We can observe that the `workflowStatus` value is set to "started" and the "progress" is set to 0.

> At this point, the recipients will receive through email or SMS a notification that a document requires to be signed by them.

Reviewer: ```suggestion
At this point, recipients will receive through email or SMS a notification that a document requires to be signed by them.
```

___

### III) Signing the document (user's view on his browser)
#### 1.1- The signer side
![schema describing the main steps of the user's point of view with lex persona](../assets/lex_persona_schema_user.png)

To summarize the main steps from the signer's point of view :
- The signer will receive an email with an invitation link to sign documents that he is concerned about OR he will be asked to click on the invitation link that we have create within our Business Applcation to sign documents.
- Once the signer has clicked on the invitation link, he will be redirected to the document page where he will need to read entirely the document in order to unlock the next step which is the signature.
- Before the signature takes part, the signer is invited to validate that he is about to sign the document that he has just read. He will need to authenticate through an Email code validation or a SMS code validation. When this step is done, he will finally be allowed to sign the document.

> - The signer will receive his copy of the signed document once all concerned parties have signed on it.

Reviewer: ```suggestion
- The signer will receive its copy of the signed document once all concerned parties have signed on it.
```

**Endpoint** :  /api/users/{{our_user_id}}/workflows

**Method** : POST

**Body Type** : JSON

**Headers** :
- Authorization : Bearer {{token}}
- Content-Type : application/json

**Payload** :
```json
{
  "name" : "Aerodynamic Silk Bottle",
  "description" : "1 rue de l'exemple, 75000 Paris",
  "steps" : [ {
    "stepType" : "signature",
    "recipients" : [ {
      "consentPageId" : "cop_NzBEGbgTkj7v5************",
      "email" : "johndoe@example.com",
      "country" : "FR",
      "phoneNumber":"+33 * ** ** ** **",
      "preferredLocale" : "fr",
      "firstName":"John",
      "lastName":"Doe"
    } ],
    "requiredRecipients" : 1,
    "validityPeriod" : 8553600000,
    "invitePeriod" : 86400000,
    "maxInvites" : 0, // this line
    "sendDownloadLink" : true,
    "allowComments" : true,
    "hideAttachments" : false,
    "hideWorkflowRecipients" : true
  } ],
  "notifiedEvents" : [ "recipientRefused", "recipientFinished", "workflowStopped", "workflowFinished" ],
  "watchers" : []
}
```

> **Once we have set our `maxInvite` to 0, we can now create the invitation link :**

Reviewer: ```suggestion
**Once we have set our `maxInvites` to 0, we can now create the invitation link :**
```
